mava/wrappers/jaxmarl.py

```diff
@@ -0,0 +1,236 @@
+
+import jax
+import jax.numpy as jnp
+from chex import Array, ArrayTree, PRNGKey
+from gymnax.environments import spaces as gymnax_spaces
+from jaxmarl.environments import spaces as jaxmarl_spaces
+from jaxmarl.environments.multi_agent_env import MultiAgentEnv
+from jumanji import specs
+from jumanji.types import StepType, TimeStep, restart
+from jumanji.wrappers import Wrapper
+
+from mava.types import Observation
+
+
+def _is_discrete(space: jaxmarl_spaces.Space) -> bool:
+    """JaxMarl sometimes uses gymnax and sometimes uses their own specs, so this is needed."""
+    return isinstance(space, (gymnax_spaces.Discrete, jaxmarl_spaces.Discrete))
+
+
+def _is_box(space: jaxmarl_spaces.Space) -> bool:
+    """JaxMarl sometimes uses gymnax and sometimes uses their own specs, so this is needed."""
+    return isinstance(space, (gymnax_spaces.Box, jaxmarl_spaces.Box))
+
+
+def _is_dict(space: jaxmarl_spaces.Space) -> bool:
+    """JaxMarl sometimes uses gymnax and sometimes uses their own specs, so this is needed."""
+    return isinstance(space, (gymnax_spaces.Dict, jaxmarl_spaces.Dict))
+
+
+def _is_tuple(space: jaxmarl_spaces.Space) -> bool:
+    return isinstance(space, (gymnax_spaces.Tuple, jaxmarl_spaces.Tuple))
+
+
+def batchify(x: Dict[str, Array], agents: List[str]) -> Array:
+    """Stack dictionary values into a single array."""
+    return jnp.stack([x[agent] for agent in agents])
+
+
+def unbatchify(x: Array, agents: List[str]) -> Dict[str, Array]:
+    """Split array into dictionary entries."""
+    return {agent: x[i] for i, agent in enumerate(agents)}
+
+
+def merge_space(
+    spec: Dict[str, Union[jaxmarl_spaces.Box, jaxmarl_spaces.Discrete]]
+) -> jaxmarl_spaces.Space:
+    """Convert a dictionary of spaces into a single space with a num_agents size first dimension.
+
+    JaxMarl uses a dictionary of specs, one per agent. For now we want this to be a single spec.
+    """
+    n_agents = len(spec)
+    single_spec = copy.deepcopy(list(spec.values())[0])
+
+    err = f"Unsupported space for merging spaces, expected Box or Discrete, got {type(single_spec)}"
+    assert _is_discrete(single_spec) or _is_box(single_spec), err
+
+    new_shape = (n_agents, *single_spec.shape)
+    single_spec.shape = new_shape
+
+    return single_spec
+
+
+def is_homogenous(env: MultiAgentEnv) -> bool:
+    """Check that all agents in an environment have the same observation and action spaces.
+
+    Note: currently this is done by checking the shape of the observation and action spaces
+    as gymnax/jaxmarl environments do not have a custom __eq__ for their specs.
+    """
+    agents = list(env.observation_spaces.keys())
+
+    main_agent_obs_shape = env.observation_space(agents[0]).shape
+    main_agent_act_shape = env.action_space(agents[0]).shape
+    # Cannot easily check low, high and n are the same, without being very messy.
+    # Unfortunately gymnax/jaxmarl doesn't have a custom __eq__ for their specs.
```

Comment:
Can you give it a try in an additional function else maybe add it as an issue

Reply:
I did try this earlier and the function just got very big and very messy :cry:

Reply:
If you can add an issue on mava then for it and we will look into it another time :pray: